{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "161b7fbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This pipeline was usen ONLY for Soil_1 to Soil_6 and Rhizosphere_1 to Rhizosphere_10 metagenomic samples#\n",
    "#The resulting contigs were used to construct the Environmental Extended Pangenome (EEP)#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c6d9b8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Host filtering for Rhizosphere samples\n",
    "#Create host genome index database\n",
    "#Solanum lycopersicum (tomato) and Arabidopsis thaliana were downloaded from NCBI and renamed#\n",
    "#Example#\n",
    "$PATH/sratoolkit.2.10.8-ubuntu64/bin/fasterq-dump SRR23012434 --split-files\n",
    "\n",
    "##Script name: make_db_tomatogenome.scr##\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "bowtie2-build $PATH/solirubrobacter/mg_hybrid_assb/tomato_genomic.fna.gz tomato_genome\n",
    "\n",
    "\n",
    "##Script name: make_db_athgenome.scr##\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "bowtie2-build $PATH/solirubrobacter/mg_hybrid_assb/athaliana_genomic.fna.gz ath_genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62aa4992",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Map tomato rhizosphere metagenomes against the host (tomato) genome with Bowtie2.\n",
    "#Align against reference genome\n",
    "##Script name: tmap_maker.scr##\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "\n",
    "SEQS=$PATH/solirubrobacter/mg_hybrid_assb\n",
    "\n",
    "\n",
    "for FAA in `ls T*.fastq.gz | perl -pe 's/\\_.*//g' | sort | uniq`\n",
    "do\n",
    "\n",
    "echo \"#!/bin/bash\" >tmap_$FAA.sh\n",
    "echo \"#$ -cwd\" >>tmap_$FAA.sh\n",
    "echo \"#$ -j y\" >>tmap_$FAA.sh\n",
    "echo \"#$ -S /bin/bash\" >>tmap_$FAA.sh\n",
    "echo \"bowtie2 -x tomato_genome -1 $SEQS/$FAA\"_R1.fastq.gz\" -2 $SEQS/$FAA\"_R2.fastq.gz\" -S $FAA.sam\"  -p 20 >>tmap_$FAA.sh\n",
    "chmod +x *.sh; done\n",
    "\n",
    "\n",
    "##Script name: amap_maker.scr##\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "\n",
    "SEQS=$PATH/solirubrobacter/mg_hybrid_assb\n",
    "\n",
    "for FAA in `ls *.fastq.gz | grep -v \"T\" | grep \"^\\R\" | perl -pe 's/\\_.*//g' | sort | uniq`\n",
    "do\n",
    "\n",
    "echo \"#!/bin/bash\" >amap_$FAA.sh\n",
    "echo \"#$ -cwd\" >>amap_$FAA.sh\n",
    "echo \"#$ -j y\" >>amap_$FAA.sh\n",
    "echo \"#$ -S /bin/bash\" >>amap_$FAA.sh\n",
    "echo \"bowtie2 -x ath_genome -1 $SEQS/$FAA\"_R1.fastq.gz\" -2 $SEQS/$FAA\"_R2.fastq.gz\" -S $FAA.sam\"  -p 20 >>amap_$FAA.sh\n",
    "chmod +x *.sh; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a989ac44",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Convert sam to bam\n",
    "##Script name: samtobam.scr ##\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "\n",
    "FILES=$PATH/solirubrobacter/mg_hybrid_assb\n",
    "\n",
    "for FAA in `ls *.sam | sed -e 's/.sam//g'`\n",
    "do\n",
    "echo \"#!/bin/bash\" >samtobam_$FAA.sh\n",
    "echo \"#$ -cwd\" >>samtobam_$FAA.sh\n",
    "echo \"#$ -j y\" >>samtobam_$FAA.sh\n",
    "echo \"#$ -S /bin/bash\" >>samtobam_$FAA.sh\n",
    "echo \"#!/bin/bash\" >>samtobam_$FAA.sh\n",
    "echo \"samtools view -bS $FILES/$FAA\".sam \"> $FAA.bam\" >>samtobam_$FAA.sh\n",
    "chmod +x *.sh; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afb9b0a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get unmapped paired sequences against tomato genome (bothEndsUnmapped)\n",
    "##Script name: getunmapped_maker.scr##\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "\n",
    "FILES=$PATH/solirubrobacter/mg_hybrid_assb\n",
    "\n",
    "for FAA in `ls *.bam | sed -e 's/.bam//g'`\n",
    "do\n",
    "\n",
    "echo \"#!/bin/bash\" >getunmapped_$FAA.sh\n",
    "echo \"#$ -cwd\" >>getunmapped_$FAA.sh\n",
    "echo \"#$ -j y\" >>getunmapped_$FAA.sh\n",
    "echo \"#$ -S /bin/bash\" >>getunmapped_$FAA.sh\n",
    "echo \"#!/bin/bash\" >>getunmapped_$FAA.sh\n",
    "\n",
    "echo \"samtools view -b -f 12 -F 256  $FILES/$FAA\".bam  \"> $FAA\"_umh_bothends.bam  >>getunmapped_$FAA.sh\n",
    "chmod +x *.sh; done\n",
    "\n",
    "#Running all the scripts#\n",
    "for N in `ls getunmapped_*.sh`; do qsub $N; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a654932a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Obtain R1 and R2 paired sequences for host filtered samples\n",
    "#sort bam file by read name (-n) to have paired reads next to each other as required by bedtools\n",
    "##Script name: sortedbam_maker.scr##\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "#!/bin/bash\n",
    "\n",
    "FILES=$PATH/solirubrobacter/mg_hybrid_assb\n",
    "\n",
    "for FAA in `ls *_umh_bothends.bam | sed -e 's/_umh_bothends.bam//g'`\n",
    "do\n",
    "\n",
    "echo \"#!/bin/bash\" >sortedbam_$FAA.sh\n",
    "echo \"#$ -cwd\" >>sortedbam_$FAA.sh\n",
    "echo \"#$ -j y\" >>sortedbam_$FAA.sh\n",
    "echo \"#$ -S /bin/bash\" >>sortedbam_$FAA.sh\n",
    "echo \"#!/bin/bash\" >>sortedbam_$FAA.sh\n",
    "\n",
    "echo \"samtools sort -n $FILES/$FAA\"_umh_bothends.bam -o \"$FAA\"_sorted.bam  >>sortedbam_$FAA.sh\n",
    "chmod +x *.sh; done\n",
    "\n",
    "#Running all the scripts#\n",
    "for N in `ls sortedbam_*.sh`; do qsub $N; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f33be2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get the filtered fastq files against host genome with bedtools\n",
    "##Script name: filth_maker.scr (Cuallicua)##\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "#!/bin/bash\n",
    "\n",
    "FILES=$PATH/solirubrobacter/mg_hybrid_assb\n",
    "BIN=$PATH/binc/bedtools2\n",
    "\n",
    "\n",
    "for FAA in `ls *_sorted.bam  | sed -e 's/_sorted.bam//g'`\n",
    "do\n",
    "echo \"#!/bin/bash\" >filth_$FAA.sh\n",
    "echo \"#$ -cwd\" >>filth_$FAA.sh\n",
    "echo \"#$ -j y\" >>filth_$FAA.sh\n",
    "echo \"#$ -S /bin/bash\" >>filth_$FAA.sh\n",
    "\n",
    "echo \"$BIN\"/bamToFastq  -i  \"$FILES/$FAA\"_sorted.bam \"-fq $FILES/$FAA\"_filtered_R1.fastq \"-fq2 $FILES/$FAA\"_filtered_R2.fastq >> filth_$FAA.sh\n",
    "chmod +x *.sh; done\n",
    "\n",
    "#Running all the scripts#\n",
    "for N in `ls filth_*.sh`; do qsub $N; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4608b78",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Quality trimming\n",
    "####Obtaining quality filtered, paired and unpaired sequences using Trimmomatic.\n",
    "##Script name: trimming_maker.scr##\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "SEQS=$PATH/solirubrobacter/mg_hybrid_assb\n",
    "BIN=$PATH/bin/Trimmomatic-0.36\n",
    "COUNT=0\n",
    "for FAA in `ls *.fastq | perl -pe 's/\\_.*//g' | sort | uniq`\n",
    "do\n",
    "echo \"#!/bin/bash\" >trim_$FAA.sh\n",
    "echo \"#$ -cwd\" >>trim_$FAA.sh\n",
    "echo \"#$ -j y\" >>trim_$FAA.sh\n",
    "echo \"#$ -S /bin/bash\" >>trim_$FAA.sh\n",
    "echo \"java -jar $BIN/trimmomatic-0.36.jar PE -threads 2 -phred33 -trimlog $SEQS/trim.log $SEQS/$FAA\"_filtered_R1.fastq\" $SEQS/$FAA\"_filtered_R2.fastq\" $SEQS/$FAA\"_paired_R1.fastq.gz\" $SEQS/$FAA\"_unpaired_R1.fastq.gz \"$SEQS/$FAA\"_paired_R2.fastq.gz\" $SEQS/$FAA\"_unpaired_R2.fastq.gz\" ILLUMINACLIP:/home/luis/bin/Trimmomatic-0.36/adapters/TruSeq3-PE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\" >>trim_$FAA.sh\n",
    "chmod +x *.sh; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99ca74ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Hybrid assembly\n",
    "#Assembly Spades\n",
    "##Script name: spades_maker.scr##\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "\n",
    "SEQS=$PATH/solirubrobacter/mg_hybrid_assb\n",
    "BIN=$PATH/binc/SPAdes-3.12.0-Linux/bin\n",
    "COUNT=0\n",
    "for FAA in `ls *_paired*.fastq.gz | perl -pe 's/\\_.*//g' | sort | uniq`\n",
    "do\n",
    "echo \"#!/bin/bash\" >$FAA.scr\n",
    "echo \"#$ -cwd\" >>$FAA.scr\n",
    "echo \"#$ -j y\" >>$FAA.scr\n",
    "echo \"#$ -S /bin/bash\" >>$FAA.scr\n",
    "echo date >> $FAA.scr\n",
    "echo  \"$BIN\"/spades.py --meta -1 \"$SEQS/$FAA\"_paired_R1.fastq.gz -2 \"$SEQS/$FAA\"_paired_R2.fastq.gz -o \"$SEQS/contigs_spades/contig_$FAA\"/ >>$FAA.scr\n",
    "echo date >>$FAA.scr\n",
    "chmod +x *.scr; done\n",
    "\n",
    "#Running all scripts#\n",
    "for N in `ls T*.scr`; do qsub -pe completenode 15 -l h_vmem=130G $N; done\n",
    "for N in `ls R*.scr`; do qsub -pe completenode 16 -l h_vmem=130G $N; done\n",
    "for N in `ls *.scr | grep -v \"_\" | grep -v \"^\\R\" | grep -v \"^\\T\" | grep -v \"m\"`; do qsub -pe completenode 15 -l h_vmem=130G $N; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2fb4900",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Creating a file with contig files from all metagenomes\n",
    "mkdir $PATH/solirubrobacter/hybrid_assb/map_spades\n",
    "for N in `ls $PATH/solirubrobacter/mg_hybrid_assb/contigs_spades/contig_*/contigs_*.fasta`; do ln -s $N $PATH/solirubrobacter/mg_hybrid_assb/map_spades/; done\n",
    "\n",
    "#Make symbolic link to paired trimmed sequences\n",
    "for N in `ls $PATH/solirubrobacter/mg_hybrid_assb/*_paired*`; do ln -s $N $PATH/solirubrobacter/mg_hybrid_assb/map_spades/; done\n",
    "\n",
    "#Map unassembled sequences in spades in order to make a second assembly using Velvet. \n",
    "##Script name: spades_map_maker.scr##\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "\n",
    "CONT=$PATH/solirubrobacter/mg_hybrid_assb/map_spades\n",
    "\n",
    "COUNT=0\n",
    "\n",
    "for FAA in `ls *paired_R1* | sed -e 's/_paired_R1.fastq.gz//g'`\n",
    "do\n",
    "let COUNT=COUNT+1\n",
    "echo \"#!/bin/bash\" >spadesmap_$FAA.scr\n",
    "echo \"#$ -cwd\" >>spadesmap_$FAA.scr\n",
    "echo \"#$ -j y\" >>spadesmap_$FAA.scr\n",
    "echo \"#$ -S /bin/bash\" >>spadesmap_$FAA.scr\n",
    "echo \"/PATH/bbmap/bbwrap.sh ref=\"$CONT\"/\"$FAA\"_contigs.fasta in=\"$CONT/$FAA\"_paired_R1.fastq.gz in2=\"$CONT/$FAA\"_paired_R2.fastq.gz build=\"$COUNT\" out=\"$FAA\".sam kfilter=22 subfilter=15 maxindel=80\" >>spadesmap_$FAA.scr\n",
    "chmod +x *.scr; done\n",
    "\n",
    "#Generating scripts#\n",
    "bash spades_map_maker.scr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eaaa12ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get Spades unmapped reads\n",
    "##Script name: get_unspa_maker.scr (Cuallicua) ##\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "\n",
    "SEQS=$PATH/solirubrobacter/mg_hybrid_assb/map_spades\n",
    "\n",
    "for FAA in `ls *.sam | sed -e 's/.sam//g'`\n",
    "do\n",
    "echo \"#!/bin/bash\" >get_un_spa_$FAA.scr\n",
    "echo \"#$ -cwd\" >>get_un_spa_$FAA.scr\n",
    "echo \"#$ -j y\" >>get_un_spa_$FAA.scr\n",
    "echo \"#$ -S /bin/bash\" >>get_un_spa_$FAA.scr\n",
    "echo \"source activate concoct_env\" >>get_un_spa_$FAA.scr\n",
    "echo \"$PATH/anaconda3/envs/concoct_env/bin/samtools view -u -f 12 -F 256 \"$SEQS/$FAA\".sam\" \"|\" \"$PATH/anaconda3/envs/concoct_env/bin/samtools bam2fq - > \"$FAA\"_ump.fastq\"  >> get_un_spa_$FAA.scr\n",
    "echo \"source deactivate\" >>get_un_spa_$FAA.scr\n",
    "chmod +x *.scr; done\n",
    "\n",
    "#Separation of unmapped reads into forward and reverse\n",
    "for FAA in `ls *_ump.fastq | sed -e 's/_ump.fastq//g'`; do \n",
    "grep '^@.*/1$' -A 3 --no-group-separator \"$FAA\"_ump.fastq > \"$FAA\"_um_r1.fastq\n",
    "grep '^@.*/2$' -A 3 --no-group-separator \"$FAA\"_ump.fastq > \"$FAA\"_um_r2.fastq; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "141fd299",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Assembly Velvet\n",
    "#Creating a symbolic link to no mapped sequences in Spades\n",
    "\n",
    "ln -s $PATH/solirubrobacter/mg_hybrid_assb/map_spades/*_um_r*.fastq $PATH/solirubrobacter/mg_hybrid_assb/velvet/\n",
    "\n",
    "#Velvet assembly of reads that did not map to Spades contigs\n",
    "##Script name: velvet_maker.scr (Cuallicua) ##\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "\n",
    "SEQS=$PATH/solirubrobacter/hybrid_assb/velvet\n",
    "BIN=$PATH/binc/velvet\n",
    "\n",
    "\n",
    "COUNT=0\n",
    "for FAA in `ls *_um_r1.fastq | perl -pe 's/\\_um_r1.fastq//g' | sort | uniq`\n",
    "do\n",
    "let COUNT=COUNT+1\n",
    "echo \"#!/bin/bash\" >velvet_$FAA.scr\n",
    "echo \"#$ -cwd\" >>velvet_$FAA.scr\n",
    "echo \"#$ -j y\" >>velvet_$FAA.scr\n",
    "echo \"#$ -S /bin/bash\" >>velvet_$FAA.scr\n",
    "\n",
    "echo \"$BIN\"/velveth \"$SEQS/$FAA\"_assemblyVelvet 31 -shortPaired -fastq \"$FAA\"_um_r1.fastq \"$FAA\"_um_r2.fastq >>velvet_$FAA.scr\n",
    "echo \"$BIN\"/velvetg \"$SEQS/$FAA\"_assemblyVelvet -exp_cov 2 -ins_length 350 >> velvet_$FAA.scr\n",
    "chmod +x *.scr; done\n",
    "\n",
    "#Running all the scripts#\n",
    "for N in `ls velvet_*`;do qsub -pe completenode 10 -l h_vmem=80G $N; done\n",
    "\n",
    "for N in `ls $PATH/solirubrobacter/mg_hybrid_assb/contigs_spades/contig_*/contigs_*.fasta`; do ln -s $N $PATH/solirubrobacter/mg_hybrid_assb/all_contigs/; done\n",
    "for N in `ls $PATH/solirubrobacter/mg_hybrid_assb/velvet/*_assemblyVelvet/*_v_*`; do ln -s $N $PATH/solirubrobacter/mg_hybrid_assb/all_contigs/; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e682b7b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Spades and Velvet contigs were merged\n",
    "#Ran in /hybrid_assb/all_contigs\n",
    "for FAA in `ls contigs_*.fasta | sed -e 's/contigs_*//g' |  sed -e 's/.fasta//g' |  grep -v \"v_\"`;\n",
    "do\n",
    "cat $PATH/solirubrobacter/mg_hybrid_assb/all_contigs/contigs_\"$FAA\".fasta $PATH/solirubrobacter/mg_hybrid_assb/all_contigs/contigs_v_\"$FAA\".fasta > $PATH/solirubrobacter/mg_hybrid_assb/assh_merge/\"$FAA\"_hb_contigs.fasta; done\n",
    "\n",
    "#Remove reads shorter than 100pb\n",
    "#Ran in /home/angelica/solirubrobacter/mg_hybrid_assb/assh_merge/\n",
    "for FAA in `ls *_hb_contigs.fasta| sed -e 's/_hb_contigs.fasta//g'`; do perl $PATH/scripts/remove_small.pl 100 \"$FAA\"_hb_contigs.fasta> \"$FAA\"_hb100_contigs.fasta; done\n",
    "\n",
    "#Hybrid contigs were indexed according to sample name\n",
    "for i in `ls *_hb100_contigs.fasta | sed -e 's/_hb100_contigs.fasta//g'`; do perl $PATH/scripts/header.fasta.numbers.pl \"$i\" \"$i\"_hb100_contigs.fasta; done"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
