{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5efc7e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Completeness evaluation for each reference genome#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b54eabfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "checkm lineage_wf $PATH/solirubrobacter/genomas_ncbi/genomes $PATH/solirubrobacter/genomas_ncbi/checkm_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd0dbdb2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02f81369",
   "metadata": {},
   "outputs": [],
   "source": [
    "#PHYLOGENOMIC ANALYSIS#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cacd6026",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Searching 16S rRNA genes using barrnap#\n",
    "#The following line was run for every genome#\n",
    "$PATH/mbin/quast-5.0.0/quast_libs/barrnap/bin/barrnap -k bac Strain1_genomic.fna > Strain1_rRNA.gff\n",
    "#Creating a file with 16S sequence for each genome#\n",
    "for N in `ls *_genomic.fna | sed s'/_genomic.fna//g'`; do $PATH/bedtools2/bin/bedtools getfasta -fi $N\"_genomic.fna\" -bed $N\"_rRNA.gff\" -fo $N\"_16S.fasta\"; done\n",
    "#The N_16S.fasta files were manually edited to remove 23S and 5S sequences#\n",
    "#Creating a file with the 16S sequences of all the genomes#\n",
    "cat *_16S.fasta > soli_outgroups_16S.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "585f472b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Creating 16S phylogeny#\n",
    "/usr/local/bin/ssu-align soli_outgroups_16S.fasta soli_outgroups_align\n",
    "/usr/local/bin/ssu-mask soli_outgroups_align/\n",
    "/usr/local/bin/ssu-mask --stk2afa soli_outgroups_align\n",
    "#The resulting *.afa file was evaluated using Seaview#\n",
    "#Creating Tree#\n",
    "$PATH/raxmlHPC-PTHREADS -T 40 -f a -x 31416 -p 31416 -# 1000 -m GTRGAMMA -s soli_outgroups_align.bacteria.afa -n 16Sout\n",
    "#The resulting tree was visualized using iTOL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58ed0f6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Analysis using Genome Similarity Score (GSS)#\n",
    "#Every sequence was renamed StrainX_sequenceY using an inhouse script. The resulted file was StrainX_edited.faa\n",
    "ls *_edited.faa > my_genomes\n",
    "\n",
    "#Running GSS scripts#\n",
    "gss_blast my_genomes\n",
    "gss_bbh my_genomes\n",
    "gss_comp my_genomes\n",
    "\n",
    "#This generates a table with similarity values for each paired comparison and a newick file with a dendrogram#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ab4a191",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Analysis using average nucleotide identity (ANI)#\n",
    "for N in `ls *.fna`; do $PATH/fastani-1.32-he1c1bb9_0/bin/fastANI -q $N --rl reference_list -o $N.ani; done\n",
    "#reference_list is a text file containing the name of the files of all genomes to be compared"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6e7e356",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Using GET_HOMOLOGUES#\n",
    "$PATH/mbin/get_homologues/get_homologues.pl -d ./proteomas -c -t0 -G\n",
    "$PATH/mbin/get_homologues/compare_clusters.pl -o sample_intersection -m -T -d $PATH/solirubrobacter/proteomas_homologues/Strain1protein_f0_0taxa_algCOG_e0_/\n",
    "$PATH/mbin/get_homologues/parse_pangenome_matrix.pl -m sample_intersection/pangenome_matrix_t0.tab -s -x\n",
    "##Using the list pangenome_matrix_t0__core_list.txt all core sequences were moved to a new directory $PATH/solirubrobacter/sample_intersection/core_seqs/##\n",
    "#Using align_maker.scr script#\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "\n",
    "for FAA in `ls *.faa | grep -v \".1.faa\" | sed 's/.faa//g'`; do\n",
    "echo \"#!/bin/bash\" >algn_$FAA.sh\n",
    "echo \"#$ -cwd\" >>algn_$FAA.sh\n",
    "echo \"#$ -j y\" >>algn_$FAA.sh\n",
    "echo \"#$ -S /bin/bash\" >>algn_$FAA.sh\n",
    "echo \"mafft $FAA.faa > \"a_\"$FAA\" >>algn_$FAA.sh\n",
    "chmod +x *.sh; done\n",
    "\n",
    "##The resulting alignments were concatennated using Seaview under the name big_align_core.aln##\n",
    "$PATH/raxmlHPC-PTHREADS -T 40 -f a -x 31416 -p 31416 -# 1000 -m PROTGAMMALG -s big_align_core.fst -n output_raxml\n",
    "#The resulting tree was visualized using iTOL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06e47ec7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "880fdd3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#CONSTRUCTION OF REFERENCE PANGENOME USING ROARY#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e75ac52",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Creating gff3 files with Prokka#\n",
    "#This was done for every genome#\n",
    "prokka --kingdom Bacteria --outdir prokka_Strain1 --genus Solirubrobacter --locustag Strain1 $PATH/solirubrobacter/genomas_ncbi/Strain1_genomic.fna\n",
    "#Running Roary#\n",
    "$PATH/bin/roary -f output -e -n -i 50 -v ./*.gff\n",
    "$PATH/anaconda3/bin/FastTreeMP -nt -gtr core_gene_alignment.aln > tree_core.nwk\n",
    "python roary_plots.py ./output_outgroup/tree_core.nwk ./output_outgroup/gene_presence_absence.csv\n",
    "Rscript create_pan_genome_plots.R\n",
    "#Translating the pangenome to upload it to KEGG#\n",
    "$PATH/transeq $PATH/solirubrobacter/pangenome/pan_genome_reference.fa $PATH/solirubrobacter/pangenome/pan_genome_reference_prot.faa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6867dda7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8889d7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Solirubrobacter RECRUITMENT FROM METAGENOMES#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "973d67d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Transforming metagenomes in fastq format to fasta##\n",
    "seqret -sequence reads.fastq -outseq reads.fasta\n",
    "\n",
    "##Renaming and changing fasta files to another directory##\n",
    "for i in *.fasta; do \n",
    "NEWNAME=\"${i/%.fastq.fasta/.fasta}\" \n",
    "mv \"$i\" \"$NEWNAME\";  done\n",
    "mv *.fasta $PATH/solurobrobacter/tblastn\n",
    "\n",
    "##Making a database for each metagenome so we can run tblastn##\n",
    "##Script name: db_maker.scr##\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "\n",
    "for FAA in `ls *.fasta | sed 's/.fasta//g'`; do\n",
    "echo \"#!/bin/bash\" >dbmk_$FAA.sh\n",
    "echo \"#$ -cwd\" >>dbmk_$FAA.sh\n",
    "echo \"#$ -j y\" >>dbmk_$FAA.sh\n",
    "echo \"#$ -S /bin/bash\" >>dbmk_$FAA.sh\n",
    "echo \"makeblastdb -in $FAA\".fasta\" -dbtype nucl\" >>dbmk_$FAA.sh\n",
    "chmod +x *.sh; done\n",
    "\n",
    "##Running tblastn using metagenomes as db and pangenome as query##\n",
    "##Script name: $PATH/solirubrobacter/tblastn/search_soli.scr##\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "\n",
    "SEQS=$PATH/solirubrobacter/tblastn\n",
    "\n",
    "for FAA in `ls *.fasta | sed 's/.fasta//g'`; do\n",
    "echo \"#!/bin/bash\" >search_$FAA.sh\n",
    "echo \"#$ -cwd\" >>search_$FAA.sh\n",
    "echo \"#$ -j y\" >>search_$FAA.sh\n",
    "echo \"#$ -S /bin/bash\" >>search_$FAA.sh\n",
    "echo \"tblastn  -query \"$SEQS\"/solirubrobacter.rep.fna -db \"$FAA\".fasta\" -out \"$FAA\".tsv -outfmt \"'\"6 qseqid qlen sseqid slen qseq evalue bitscore pident length mismatch gapopen qstart qend sstart send\"'\" >>search_$FAA.sh\n",
    "chmod +x *.sh; done\n",
    "\n",
    "bash search_soli.scr\n",
    "#Running all the scripts##\n",
    "for N in `ls search_*.sh`; do qsub $N; done \n",
    "\n",
    "##This generates files with the reads in each metagenome that align to the pangenome of Solirubrobacter##\n",
    "##Filtering results according to bitscore value##\n",
    "for FAA in `ls *.tsv | sed -e 's/.tsv//g'`; do cat \"$FAA\".tsv | perl -pe '$name_col=0; $score_col=11; while(<>) { s/\\r?\\n//; @F=split /\\t/, $_; ($n, $s) = @F[$name_col, $score_col]; if (! exists($max{$n})) { push @names, $n }; if (! exists($max{$n}) || $s > $max{$n}) { $max{$n} = $s; $best{$n} = () }; if ($s == $max{$n}) { $best{$n} .= \"$_\\n\" }; } for $n (@names) { print $best{$n} } ' > best; perl -e ' $column=0; $unique=0; while(<>) { s/\\r?\\n//; @F=split /\\t/, $_; if (! ($save{$F[$column]}++)) { print \"$_\\n\"; $unique++ } } ' best > \"$FAA\"_best_uniq; rm best; done\n",
    "\n",
    "#Filter in order to get only 60% identity\n",
    "for N in `ls *_best_uniq| sed -e 's/\\_best_uniq//g'`; do awk -F \"\\t\" '{if ($3>60) print $0}' \"$N\"_best_uniq > \"$N\"_pident_m5nr.bout;done\n",
    "\n",
    "#Simplify output of best hits files\n",
    "for i in `ls *_pident_best_uniq| sed -e 's/\\_pident_best_uniq//g'`; do awk '{print $1\"\\t\"$3\"\\t\"$11\"\\t\"$12\"\\t\"$2}' \"$i\"_pident_best_uniq > \"$i\"_best.simple.tsv; done\n",
    "\n",
    "##Extract metagenomic reads and create a fasta file for assemble##\n",
    "##Extracting subject sequence ids##\n",
    "for N in `ls *_best_uniq | sed 's/_best_uniq//g'`; do cut -f3 $N\"_best_uniq\" > $N\"_id.txt\"; done \n",
    "\n",
    "##Extracting subject sequences from fasta files##\n",
    "##Script name: filter_soli.scr##\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "\n",
    "SEQS=$PATH/solirubrobacter/metagenomes\n",
    "\n",
    "for FAA in `ls *.fasta | sed 's/.fasta//g'`; do\n",
    "echo \"#!/bin/bash\" >filter_$FAA.sh\n",
    "echo \"#$ -cwd\" >>filter_$FAA.sh\n",
    "echo \"#$ -j y\" >>filter_$FAA.sh\n",
    "echo \"#$ -S /bin/bash\" >>filter_$FAA.sh\n",
    "echo \"/home/qiime/bin/filter_fasta.py -f $SEQS/$FAA\".fastq\" -o $FAA\"_soli.fastq\" -s $FAA\"_id.txt\"\" >>filter_$FAA.sh\n",
    "chmod +x *.sh; done\n",
    "\n",
    "bash filter_soli.scr\n",
    "\n",
    "#Running all scripts#\n",
    "for N in `ls filter_*.sh`; do qsub $N; done \n",
    "\n",
    "##The resulting fasta files contain only sequences from Solirubrobacter##\n",
    "##Sequences of Solirubrobacter: Renaming fastq files##\n",
    "\n",
    "for i in *_R1_soli.fastq; do \n",
    "NEWNAME=\"${i/%_R1_soli.fastq/_soli_R1.fastq}\" \n",
    "mv \"$i\" \"$NEWNAME\";  done\n",
    "\n",
    "for i in *_R2_soli.fastq; do \n",
    "NEWNAME=\"${i/%_R2_soli.fastq/_soli_R2.fastq}\" \n",
    "mv \"$i\" \"$NEWNAME\";  done\n",
    "\n",
    "\n",
    "#Obtaining quality filtered, paired and unpaired Solirubrobacter sequences using Trimmomatic.\n",
    "##Script name: trimm_maker.scr##\n",
    "##This is run in tblastn directory, where the fastq files with Solirubrobacter reads are found ##\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "\n",
    "FILES=$PATH/solirubrobacter/tblastn\n",
    "SEQS=$PATH/solirubrobacter/trimming\n",
    "BIN=$PATH/bin/Trimmomatic-0.36\n",
    "\n",
    "COUNT=0\n",
    "for FAA in `ls *.fastq| perl -pe 's/\\_.*//g' | sort | uniq`\n",
    "do\n",
    "echo \"#!/bin/bash\" >trim_$FAA.sh\n",
    "echo \"#$ -cwd\" >>trim_$FAA.sh\n",
    "echo \"#$ -j y\" >>trim_$FAA.sh\n",
    "echo \"#$ -S /bin/bash\" >>trim_$FAA.sh\n",
    "echo \"java -jar $BIN/trimmomatic-0.36.jar PE -threads 2 -phred33 -trimlog $SEQS/trim.log $FILES/$FAA\"_soli_R1.fastq\" $FILES/$FAA\"_soli_R2.fastq\" $SEQS/$FAA\"_s_paired_R1.fastq.gz\" $SEQS/$FAA\"_s_unpaired_R1.fastq.gz\" $SEQS/$FAA\"_s_paired_R2.fastq.gz\" $SEQS/$FAA\"_s_unpaired_R2.fastq.gz\" ILLUMINACLIP:/home/luis/bin/Trimmomatic-0.36/adapters/TruSeq3-PE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\" >>trim_$FAA.sh\n",
    "chmod +x *.sh; done\n",
    "\n",
    "#Running all the scripts#\n",
    "for N in `ls trim_*.sh`; do qsub $N; done\n",
    "\n",
    "#Converting fastq files to fasta#\n",
    "#Script name convert_fastqfasta.scr in $PATH/solirubrobacter/trimming#\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "\n",
    "SEQS=$PATH/solirubrobacter/trimming\n",
    "\n",
    "COUNT=0\n",
    "for FAA in `ls *_s_paired_R*.fastq.gz | sed 's/.fastq.gz//g'`\n",
    "do\n",
    "echo \"#!/bin/bash\" >conv_$FAA.sh\n",
    "echo \"#$ -cwd\" >>conv_$FAA.sh\n",
    "echo \"#$ -j y\" >>conv_$FAA.sh\n",
    "echo \"#$ -S /bin/bash\" >>conv_$FAA.sh\n",
    "echo \"gunzip $SEQS/$FAA\".fastq.gz\"\" >>conv_$FAA.sh\n",
    "echo \"/usr/bin/seqret -sequence $SEQS/$FAA\".fastq\" -outseq $SEQS/$FAA\".fasta\"\" >>conv_$FAA.sh\n",
    "echo \"gzip $SEQS/$FAA\".fastq\"\" >>conv_$FAA.sh\n",
    "chmod +x *.sh; done\n",
    "\n",
    "\n",
    "bash convert_fastqfasta.scr\n",
    "\n",
    "#Running all the scripts#\n",
    "for N in `ls conv_*R*.sh`; do qsub $N; done\n",
    "\n",
    "#Concatenating paired reads#\n",
    "for N in `ls *_s_paired_R*.fasta | sed 's/_s_paired_R1.fasta//g' | sed 's/_s_paired_R2.fasta//g'`; do cat $N\"_s_paired_R1.fasta\" $N\"_s_paired_R2.fasta\" > $N\"_s_reads.fasta\"; done\n",
    "\n",
    "#Files *_s_reads.fasta contain reads of Solirubrobacter. There’s one file per metagenome#\n",
    "#Collapsing fasta files according to environment#\n",
    "cat r*_s_reads.fasta > rhizo_s_reads.fasta\n",
    "cat Soil_1_s_reads.fasta Soil_2_s_reads.fasta Soil_3_s_reads.fasta Soil_4_s_reads.fasta Soil_5_s_reads.fasta Soil_6_s_reads.fasta > soil_s_reads.fasta\n",
    "#The same was carried out with lakes, phyllospheres, foreign soils, foreign rhizospheres, and sediments#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1e093e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Finding Solirubrobacter in metagenomes#\n",
    "#Running promer for every reference using each metagenome Solirubrobacter reads as query#\n",
    "for N in `ls soil*_s_reads.fasta | sed 's/_s_reads.fasta//g'`; do\n",
    "promer pan_genome_reference.fa $N\"_s_reads.fasta\" \n",
    "mv out.delta \"pangenome_\"$N\".delta\"; done\n",
    "#To edit promer graphs\n",
    "for N in `ls pangenome_soil_*.delta | sed 's/.delta//g'`; do\n",
    "show-coords -c -d -k -l -r -T $N\".delta\" > $N\".show-coords.tsv\"\n",
    "python3 promer_deid.py $N\".show-coords.tsv\" $N;\n",
    "done\n",
    "#promer_deid.py it's an inhouse script#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "439c99f0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3c3b328",
   "metadata": {},
   "outputs": [],
   "source": [
    "#CREATING ENVIRONMENTAL EXTENDED PANGENOME (EEP)#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa81d2a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Search for Solirubrobacter contigs that contain a reference sequence and adjacently a gene not found int he reference pangenome#\n",
    "#Using the script bbmap_maker.scr\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "\n",
    "COUNT=0\n",
    "\n",
    "for FAA in `ls *_hb100_contigs.fasta.numbered.fas | sed -e 's/_hb100_contigs.fasta.numbered.fas//g'`; do\n",
    "let COUNT=COUNT+1\n",
    "echo \"#!/bin/bash\" >bbmap_$FAA.scr\n",
    "echo \"#$ -cwd\" >>bbmap_$FAA.scr\n",
    "echo \"#$ -j y\" >>bbmap_$FAA.scr\n",
    "echo \"#$ -S /bin/bash\" >>bbmap_$FAA.scr\n",
    "echo \"$PATH/binc/bbmap/bbmap.sh ref=$FAA\"_hb100_contigs.fasta.numbered.fas\" nodisk in=pan_genome_reference.fa covstats=$FAA\"_p_constats.txt\" covhist=$FAA\"_p_covhist.txt\" basecov=$FAA\"_p_basecov.txt\" bincov=$FAA\"_p_bincov.txt\" outm=$FAA\"_m.bam\" outu=$FAA\"_um.bam\" k=14 minid=0.6\" >> bbmap_$FAA.scr\n",
    "\n",
    "chmod +x *.scr; done\n",
    "\n",
    "#Retrieve contigs’ id’s and sequences\n",
    "\n",
    "for N in `ls *_m.sam |sed 's/_m.sam//g' `; do grep \"^S\" $N\"_m.sam\" | awk '{print $4, $5}' > $N\"_scontigs_id.txt\"; done\n",
    "for N in `ls *_hb100_contigs.fasta.numbered.fas |sed 's/_hb100_contigs.fasta.numbered.fas//g' `; do grep -A1 -Fwf $N\"_scontigs_id.txt\" $N\"_hb100_contigs.fasta.numbered.fas\" > $N\"_scontigs.fasta\"; done\n",
    "\n",
    "#Filtering the selected contigs\n",
    "\n",
    "makeblastdb -in $PATH/solirubrobacter/contig_search/pan_genome_reference_prot.faa -dbtype prot \n",
    "\n",
    "#For false_postivies.scr in metagenome soli contigs: \n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "\n",
    "SEQS=$PATH/solirubrobacter/contig_search\n",
    "COUNT=0\n",
    "\n",
    "for FAA in `ls *_scontigs.fasta | sed 's/_scontigs.fasta//g'`; do\n",
    "let COUNT=COUNT+1\n",
    "echo \"#!/bin/bash\" >fp_$FAA.scr\n",
    "echo \"#$ -cwd\" >>fp_$FAA.scr\n",
    "echo \"#$ -j y\" >>fp_$FAA.scr\n",
    "echo \"#$ -S /bin/bash\" >>fp_$FAA.scr\n",
    "echo \"blastx -db $SEQS/pan_genome_reference_prot.faa  -query $SEQS/$FAA\"_scontigs.fasta\" -outfmt 7 -evalue 1e-5 -num_alignments 500 -num_threads 4 -out $FAA\"_falsepos.bout\"\" >> fp_$FAA.scr\n",
    "\n",
    "chmod +x *.scr; done\n",
    "\n",
    "for N in `ls *_falsepos.bout`; do grep -cw \"0 hits found\" $N; done \n",
    "for N in `ls *_falsepos.bout | sed 's/_falsepos.bout//g'`; do grep -w -B2 \"0 hits found\" $N\"_falsepos.bout\" | grep \"# Query: \" | sed 's/# Query: //g' > $N\"_id_contig_REMOVE.txt\"; done  \n",
    "\n",
    "#This generates a file with the names of the contigs that were false positives and should be removed from the new pangenome#\n",
    "\n",
    "#Removing contigs that did not pass the filtering step\n",
    "for N in `ls *_scontigs.fasta | sed 's/_scontigs.fasta//g'`; do /home/qiime/bin/filter_fasta.py -f $N\"_scontigs.fasta\" -o $N\"_scontigs_filt.fasta\" -n -s $N\"_id_contig_REMOVE.txt\"; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19b6b84b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Renaming sequences using the script header.fasta.numbers.pl (rename_seq.scr)#\n",
    "for genome in `ls *_genomic.fna | sed 's/_genomic.fna//g'`; do perl $PATH/scripts/header.fasta.numbers.pl $genome \"$genome\"_genomic.fna; done\n",
    "for genome in `ls *_genomic.fna.numbered.fas | sed 's/_genomic.fna.numbered.fas//g'`; do awk '{if ($1~\">\"){print $1} else {(\":\");print}}' $genome\"_genomic.fna.numbered.fas\" > $genome\"_scontigs.fasta\"; done\n",
    "\n",
    "#A symbolic link for all *numbered* files from Solirubrobacter was created in the pangenome_new directory in Cuallicua#\n",
    "\n",
    "ln -s $PATH/solirubrobacter/genomas_ncbi/*scontigs.fasta ./\n",
    "ln -s $PATH/solirubrobacter/pangenome/S*.fasta ./\n",
    "\n",
    "for N in `ls T*_scontigs.fasta | sed 's/_scontigs.fasta//g'`; do $PATH/anaconda3/bin/prodigal -i $N\"_scontigs.fasta\" -o $N\"_sgenes.fasta\" -a $N\"_sproteins.faa\" -p meta; done \n",
    "\n",
    "#Concatenate all genomes with .fasta extension and modified headers#\n",
    "cat *.faa Ssoli.fasta Staibaiensis.fasta SpURHD.fasta SpCPCC.fasta Spauli.fasta Sginsenosidimutans.fasta Sphytolaccae.fasta > solirubrobacter_new_filt.faa\n",
    "\n",
    "#Clustering using CD-HIT at 50% identity using the script cdhit_50.scr#\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "cd-hit -i solirubrobacter_new.faa -o soli_new_50 -n 3 -c 0.50 -aS 0.7 -d 0  -M 30000 -T 10 >  soli_new_50_filt.cdhit.out\n",
    "\n",
    "\n",
    "#Converting soli_50.clstr file for QIIME#\n",
    "perl -pne 's/\\t//g;s/^.*,//g;s/\\.\\.\\..*$//g;s/\\n/\\t/g;s/\\>Cluster\\ /\\n/g;s/\\>//g; eof && do{chomp; print \"$_ \\n\"; exit}' soli_new_50.clstr > solirubrobacter_new_filt.otu\n",
    "\n",
    "#Extracting representative sequence from each cluster# \n",
    "pick_rep_set.py -i solirubrobacter_new_filt.otu -f solirubrobacter_new.faa -o solirubrobacter_new.rep.fna\n",
    "grep -c \">\" solirubrobacter_new_filt.rep.fna\n",
    "\n",
    "#Obtaining ID from the representative sequences#\n",
    "grep \">\" solirubrobacter_new_filt.rep.fna > identif_new_soli\n",
    "awk '{if ($1~\">\"){print $2} else {(\":\");print}}' identif_new_soli > identif_new_edit\n",
    "sed 's/^/>/g' identif_new_edit > id_new_edit\n",
    "\n",
    "#Generating count table using QIIME with the script count_table.scr#\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "/home/qiime/bin/make_otu_table.py -i solirubrobacter_new_filt.otu -o solirubrobacter_new_filt.biom\n",
    "/home/qiime/bin/biom convert --to-tsv -i solirubrobacter_new_filt.biom -o solirubrobacter_new_filt.tsv\n",
    "\n",
    "#File solirubrobacter_new.tsv is further analyzed using R. PLEASE REFER TO THE CORRESPONDING FILE #"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "352353c8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2210be23",
   "metadata": {},
   "outputs": [],
   "source": [
    "#GENE FAMILIES SEARCH#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbcc592e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Search for gene families#\n",
    "#The file solirubrobacter_new_filt.otu was duplicated and renamed solirubrobacter_clsr.otu\n",
    "sed -i 's/_//g' solirubrobacter_clsr.otu\n",
    "cut -f1 solirubrobacter_clsr.otu > clstr_id \n",
    "sed -i 's/[0-9]//g' solirubrobacter_clsr.otu\n",
    "paste -d\"\\t\" clstr_id solirubrobacter_clsr.otu > solirubrobacter_clstr.otu\n",
    "sed -i '1d' solirubrobacter_clstr.otu\n",
    "sed -i '1d' clstr_id\n",
    "\n",
    "#This file contains the name of the metagenome/genome that has a sequence of the cluster (first column id of the file)\n",
    "rm solirubrobacter_clsr.otu\n",
    "\n",
    "#Getting counts per cluster\n",
    "for N in `cat lista`; do awk -F' ' '{ count=0; for (i=1; i<=NF; i++) if ($i == \"$N\") count++; print count }' solirubrobacter_clstr.otu > $N.genefam.counts.txt; done\n",
    "#Adding cluster number to each file\n",
    "for N in `ls *.genefam.counts.txt | sed 's/.genefam.counts.txt//g'`; do paste -d\"\\t\" clstr_id $N.genefam.counts.txt > $N.genefam.counts.tsv; done\n",
    "rm *.genefam.counts.txt\n",
    "\n",
    "#Adding headers to the columns\n",
    "for N in `ls *.genefam.counts.tsv | sed 's/.genefam.counts.tsv//g'`; \n",
    "do \n",
    "awk -F \"\\t\" 'BEGIN {print \"Cluster Counts\"} {print $0}' $N.genefam.counts.tsv > $N.tmp\n",
    "    mv $N.tmp $N.genefam.counts.tsv;\n",
    "done\n",
    "\n",
    "for N in `ls S*genefam.counts.tsv | sed 's/.genefam.counts.tsv//g'`; do sed -i 's/Counts/'$N'/g' $N.genefam.counts.tsv; done\n",
    "\n",
    "#Obtaining annotation from the pangenome’s annotation\n",
    "for N in `ls gene_*.tsv | sed 's/gene_//g' | sed 's/.tsv//g'`; \n",
    "do  \n",
    "cut -f1 $PATH/Drive/Solirubrobacter/gene_fam/\"gene_\"$N.tsv > $PATH/Drive/Solirubrobacter/gene_fam/\"id_\"$N\n",
    "sed -i 's/^/RepSeq_/g' $PATH/Drive/Solirubrobacter/gene_fam/\"id_\"$N\n",
    "sed -i 's/$/_/g' $PATH/Drive/Solirubrobacter/gene_fam/\"id_\"$N\n",
    "grep -Ff $PATH/Drive/Solirubrobacter/gene_fam/\"id_\"$N $PATH/Drive/Solirubrobacter/pangenome_cuallicua/pangenome_new/filtered/pangenome_annot.tsv > $PATH/Drive/Solirubrobacter/gene_fam/\"gene\"$N\"_annot.tsv\";\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c0c79da",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "567574a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#SEARCH FOR PLANT-MICROORGANISM INTERCATION GENES#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75ad136c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Using HMM for Actinobacteria plant-microorganism interaction genes#\n",
    "#For every file a similar script was run# \n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "$PATH/anaconda3/bin/hmmsearch -A 3 $PATH/solirubrobacter/hmm_databases/orthogroups_database_actinobacteria_one.hmm $PATH/solirubrobacter/hmm_databases/soils_s_reads.fasta > $PATH/solirubrobacter/hmm_databases/soils_hmm/soils_pm\n",
    "\n",
    "#Renaming file\n",
    "mv 3 results_one_soils\n",
    "\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "$PATH/anaconda3/bin/hmmsearch -A 3 $PATH/solirubrobacter/hmm_databases/orthogroups_database_actinobacteria_two.hmm $PATH/solirubrobacter/hmm_databases/soils_s_reads.fasta > $PATH/solirubrobacter/hmm_databases/soils_hmm/soils_pm2\n",
    "\n",
    "#Renaming file\n",
    "mv 3 results_two_soils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fba27448",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd7264f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#SEARCH FOR SULFUR METABOLISM PROTEINS#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed78ab5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Creating a db from contigs of all metagenomes using script db_maker.scr#\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "\n",
    "for FAA in `ls *.fasta.numbered.fas | sed 's/.fasta.numbered.fas//g'`; do\n",
    "echo \"#!/bin/bash\" >dbmk_$FAA.sh\n",
    "echo \"#$ -cwd\" >>dbmk_$FAA.sh\n",
    "echo \"#$ -j y\" >>dbmk_$FAA.sh\n",
    "echo \"#$ -S /bin/bash\" >>dbmk_$FAA.sh\n",
    "echo \"makeblastdb -in $FAA\".fasta.numbered.fas\" -dbtype nucl\" >>dbmk_$FAA.sh\n",
    "chmod +x *.sh; done\n",
    "\n",
    "##Running tblastn using metagenomes contigs as db and sulfur proteins as query ##\n",
    "##Sulfur proteins NCBI Reference Sequences: NP_417244.1 and WP_012265056.1##\n",
    "##Script name: search_sulph.scr ##\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "\n",
    "SEQS=$PATH/solirubrobacter/mg_hybrid_assb/assh_merge\n",
    "\n",
    "for FAA in `ls *.fasta.numbered.fas | sed 's/.fasta.numbered.fas//g'`; do\n",
    "echo \"#!/bin/bash\" >search_$FAA.sh\n",
    "echo \"#$ -cwd\" >>search_$FAA.sh\n",
    "echo \"#$ -j y\" >>search_$FAA.sh\n",
    "echo \"#$ -S /bin/bash\" >>search_$FAA.sh\n",
    "echo \"tblastn  -query \"$SEQS\"/sulph_prot_miss.fasta -db \"$FAA\".fasta.numbered.fas\" -out \"$FAA\".tsv -outfmt \"'\"6 qseqid qlen sseqid slen qseq evalue bitscore pident length mismatch gapopen qstart qend sstart send\"'\" >>search_$FAA.sh\n",
    "chmod +x *.sh; done\n",
    "\n",
    "bash search_sulph.scr\n",
    "\n",
    "for N in `ls search_*.sh`; do qsub $N; done \n",
    "\n",
    "##This generates files with the sequence in each metagenome that aligns to the pangenome of Solirubrobacter##\n",
    "##Filtering results according to bitscore value##\n",
    "\n",
    "for FAA in `ls *_hb100_contigs.tsv | sed -e 's/_hb100_contigs.tsv//g'`; do cat $FAA\"_hb100_contigs.tsv\" | perl -pe '$name_col=0; $score_col=11; while(<>) { s/\\r?\\n//; @F=split /\\t/, $_; ($n, $s) = @F[$name_col, $score_col]; if (! exists($max{$n})) { push @names, $n }; if (! exists($max{$n}) || $s > $max{$n}) { $max{$n} = $s; $best{$n} = () }; if ($s == $max{$n}) { $best{$n} .= \"$_\\n\" }; } for $n (@names) { print $best{$n} } ' > best; perl -e ' $column=0; $unique=0; while(<>) { s/\\r?\\n//; @F=split /\\t/, $_; if (! ($save{$F[$column]}++)) { print \"$_\\n\"; $unique++ } } ' best > \"$FAA\"_best_uniq; rm best; done\n",
    "\n",
    "\n",
    "\n",
    "#Filter in order to get only 60% identity\n",
    "for N in `ls *_best_uniq | sed -e 's/\\_best_uniq//g'`; do awk -F \"\\t\" '{if ($8>50) print $0}' \"$N\"_best_uniq > \"$N\"_pident.bout; done\n",
    "\n",
    "#Simplify output of best hits files\n",
    "for i in `ls *_pident_best_uniq | sed -e 's/\\_pident_best_uniq//g'`; do awk '{print $1\"\\t\"$3\"\\t\"$11\"\\t\"$12\"\\t\"$2}' \"$i\"_pident_best_uniq > \"$i\"_best.simple.tsv; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd7a7f85",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b945fa60",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4201e2fa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba69cbf0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
